### PR TITLE
Add file context resource

### PR DESCRIPTION
**Pull request by joerg**
Basically this is a copy of the port provider for managing file resources. I had to create it to get mysql running which is why I added it as an example. The noregex_filespec for sure needs to be expanded but it fits all examples I have found so far.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -102,6 +102,38 @@ allow openvpn_t openvpn_etc_t:file { write unlink };
 end
 ```
 
+### fcontext
+
+Allows managing the SELinux context of files. This may be needed to allow certain daemons access to their files and similar.
```
Comment (BackSlasher) [inline README.md]: Please rephrase:
```text
Allows managing the SELinux context of files.  
This can be used to grant SELinux-protected daemons access to additional / moved files
```

File: providers/fcontext.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+def whyrun_supported?
+  true
+end
+
+use_inline_resources
+
+# Create if doesn't exist, do not touch if fcontext is already registered
+action :add do
+  escaped_file_spec = Regexp.escape(new_resource.file_spec)
+  noregex_file_spec = new_resource.file_spec.split(/(\(\/\.\*\)\?|\.\*)$/)[0]
```
Comment (BackSlasher) [inline providers/fcontext.rb]: I saw I have `find -regex` on my machine.
I think we can use something like `find $(<safe path you extracted>) -regex $(entire expression)`
I'm just afraid of applying `restorecon` on other files and causing side effects.
Your opinion?

**Comment by BackSlasher** [comment]
Thanks for your contribution! I added 2 comments, comment / modify.

**Comment by BackSlasher** [comment]
@joerg I made a new minor version. Please rebase and add `use_selinux` wrapper

File: providers/fcontext.rb (hunk 2 above)
Comment (joerg) [inline providers/fcontext.rb]: Hmm, I get what you mean, but restorecon just applies what is set in selinux. I don't have too much experience with selinux, but do you thing it is that evil?

Comment (BackSlasher) [inline providers/fcontext.rb]: I have little exp too :(
Let's keep it like that and add `#XXX: Might run on unrelated files` above
it.
BTW, I think that you can get the noregex by escaping and then splitting
before the first '\' (which would indicate a regex), and adding a `*` at
the end.

*Nitzan Raz*
http://backslasher.net

Comment (joerg) [inline providers/fcontext.rb]: Ok, I'll do all of that tomorrow.
And splitting at \ wont work. Look at the examples I gave for mysql. Pretty much all regexes I found (semanage fcontext -l) are either .* or (/.*)? at the end. But looking at the output of this command I get the impression that it would be best to explicitly pass the restorecon pathname as an attribute. What do you think?

Comment (BackSlasher) [inline providers/fcontext.rb]: Your regex works (while mine doesn't) but I don't understand the expression. Could you explain it please?
We can keep this and if I'll find something better (which also works) I'll patch this

Comment (BackSlasher) [inline providers/fcontext.rb]: Oh and asking for the `restorecon` sounds like a last resort for me because a human would be able to deduce it from the pattern specified.

Comment (joerg) [inline providers/fcontext.rb]: Well, my regexp simply removes a ending '(/.*)?' or '.*'. Remove all the escapes (backslashes) and you will see.
How about that: We add restorecon as an optional parameter and extend the current regexp by other patterns from "semanage fcontext -l". If we make restorecon a separate provider we could even produce nice log output so people can see, and debug, what goes wrong.
Is have some spare time tomorrow and could implement that. Getting the no_regex_filename would then of course be a library definition.

Comment (joerg) [inline providers/fcontext.rb]: Actually, thinking about it we should just apply restorecon recusively to the File.dirname of the filespec. Its kind of evil, but way less magic.

Comment (joerg) [inline providers/fcontext.rb]: Ok, there is no way to properly support _ALL_ regexes. Some examples from fcontext -l:
* /var/lib/amanda/[^/]+/index(/.*)?
* /var/mailman/pythonlib(/.*)?/.+\.so(\..*)?
* /usr/lib(64)?/vmware/lib(/.*)?/HConfig\.so

So I came up with a simple and transparent solution. Tell me what you think about it and if it is ok, I will instantly implement it and send you a pull request.
```ruby
def get_restorecon_path (regex_path)
  path = regex.path.sub(/\\/,'') # Remove backslashes
  return path if File.exist?(path) # If regular file, just return
  path.count('/').times do
    path = File.dirname(path) # Splits at last '/' and returns front part
    return path if File.directory?(path) # Return is path is a valid directory
  end
  return '/' # This actually should never happen.
end
```
In the end we will just use restorecon -R #{get_restorecon_path(new_resource.file_spec)}

What do you think...

Comment (BackSlasher) [inline providers/fcontext.rb]: Your idea is fine, but it seems a lot of fuss (also - making convergence decisions based on files present during compilation seems fishy).

I'd really like to use find with regex (try `find -regex '/usr/lib(64)?/vmware/lib(/.*)?/HConfig.so'`). Problems are:

1. `find` searches the entire FS first (I get "access denied" on files that don't even resemble the path) - should be mitigated by extracting the safest directory
2. There are several regex types (as indicated by `-regextype` in the manpage). I couldn't find the type `restorecon` uses, and I'm wary of discovering later that we've used the wrong one.

I admit issue 1 is the solution we're trying to find anyway, though :)
I think our issue is extracting a path from the regex, doesn't have to be the narrowest because we'll additionally use the regex thing. We can give up on it completely too, I guess.
Opinions?

Comment (BackSlasher) [inline providers/fcontext.rb]: http://unix.stackexchange.com/questions/187660/what-is-selinuxs-regex-type

Comment (joerg) [inline providers/fcontext.rb]: I wouldn't execute this definition during compilation. It should be lazily evaluated at run time.
Well, I looked through some of the selinux sources and I think its save to assume they use basic POSIX regexes.
We could narrow the find down by using the get_restorecon_path I suggested above. That should definitely speed things up a lot. That should also remove most "access denied" problems.

Giving up is not really an option in my eyes. We do have solutions, we just don't like them. ;)

Comment (BackSlasher) [inline providers/fcontext.rb]: OK, so:

1. get_restorecon_path returns the narrowest directory - this is fine. You
have a typo in comment (not forward part, backwards part).
  I know this is the nth time I'm asking, but can it be done using regex
trickery instead of checking the actual FS? I thought so but found out that
it escapes `-`, while `'a-b'[/-.*/]` works, which makes it nontrivial.
2. We use your function to determine the directory, and invoke something
like `find $BASEDIR -regex $REGEX -exec restorecon -R "{}" \;`

Good plan?

*Nitzan Raz*
http://backslasher.net

Comment (joerg) [inline providers/fcontext.rb]: 1: This is more of less pseudo-code. I somehow tested it, and it will work, but you know...
How could you ever expand a .* without checking the actual FS?
2: If we'd use -R we could drop the find all along. ;)

I think we have a plan. I'll try to get that ready within the next few hours.

Comment (BackSlasher) [inline providers/fcontext.rb]: We'll avoid having to expand it because we'll pass it to `find` (as the
regex), which will expand it by itself.
But I think I obsessed over it too much. Let's do whatever and if someone
complains we'll invest the time in fixing it. Please keep a `#XXX` above it
though :)

*Nitzan Raz*
http://backslasher.net
